1. Data Exploration and Preprocessing
●	Begin by loading and exploring the "Alphabets_data.csv" dataset. Summarize its key features such as the number of samples, features, and classes.
●	Execute necessary data preprocessing steps including data normalization, managing missing values.


In [29]:
import pandas as pd
import warnings
warnings.filterwarnings('ignore')
data=pd.read_csv('sonardataset.csv')


In [3]:
data

x_1     x_2     x_3     x_4     x_5     x_6     x_7     x_8     x_9  \
0    0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1    0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2    0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3    0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4    0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   
..      ...     ...     ...     ...     ...     ...     ...     ...     ...   
203  0.0187  0.0346  0.0168  0.0177  0.0393  0.1630  0.2028  0.1694  0.2328   
204  0.0323  0.0101  0.0298  0.0564  0.0760  0.0958  0.0990  0.1018  0.1030   
205  0.0522  0.0437  0.0180  0.0292  0.0351  0.1171  0.1257  0.1178  0.1258   
206  0.0303  0.0353  0.0490  0.0608  0.0167  0.1354  0.1465  0.1123  0.1945   
207  0.0260  0.0363  0.0136  0.0272  0.0214  0.0338  0.0655  0.1400  0.1843   

       x_10  ...    x_52    x_53    x_54    x_55    x_56    x_57    x_58  \
0    0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084   
1    0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
2    0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   
3    0.1264  ...  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050  0.0044   
4    0.4459  ...  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072  0.0048   
..      ...  ...     ...     ...     ...     ...     ...     ...     ...   
203  0.2684  ...  0.0116  0.0098  0.0199  0.0033  0.0101  0.0065  0.0115   
204  0.2154  ...  0.0061  0.0093  0.0135  0.0063  0.0063  0.0034  0.0032   
205  0.2529  ...  0.0160  0.0029  0.0051  0.0062  0.0089  0.0140  0.0138   
206  0.2354  ...  0.0086  0.0046  0.0126  0.0036  0.0035  0.0034  0.0079   
207  0.2354  ...  0.0146  0.0129  0.0047  0.0039  0.0061  0.0040  0.0036   

       x_59    x_60  Y  
0    0.0090  0.0032  R  
1    0.0052  0.0044  R  
2    0.0095  0.0078  R  
3    0.0040  0.0117  R  
4    0.0107  0.0094  R  
..      ...     ... ..  
203  0.0193  0.0157  M  
204  0.0062  0.0067  M  
205  0.0077  0.0031  M  
206  0.0036  0.0048  M  
207  0.0061  0.0115  M  

[208 rows x 61 columns]

In [14]:
data.Y.value_counts()

Y
M    111
R     97
Name: count, dtype: int64

In [4]:
data.shape

(208, 61)

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 208 entries, 0 to 207
Data columns (total 61 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   x_1     208 non-null    float64
 1   x_2     208 non-null    float64
 2   x_3     208 non-null    float64
 3   x_4     208 non-null    float64
 4   x_5     208 non-null    float64
 5   x_6     208 non-null    float64
 6   x_7     208 non-null    float64
 7   x_8     208 non-null    float64
 8   x_9     208 non-null    float64
 9   x_10    208 non-null    float64
 10  x_11    208 non-null    float64
 11  x_12    208 non-null    float64
 12  x_13    208 non-null    float64
 13  x_14    208 non-null    float64
 14  x_15    208 non-null    float64
 15  x_16    208 non-null    float64
 16  x_17    208 non-null    float64
 17  x_18    208 non-null    float64
 18  x_19    208 non-null    float64
 19  x_20    208 non-null    float64
 20  x_21    208 non-null    float64
 21  x_22    208 non-null    float64
 22  x_

In [ ]:
data.describe()

x_1         x_2         x_3         x_4         x_5         x_6  \
count  208.000000  208.000000  208.000000  208.000000  208.000000  208.000000   
mean     0.029164    0.038437    0.043832    0.053892    0.075202    0.104570   
std      0.022991    0.032960    0.038428    0.046528    0.055552    0.059105   
min      0.001500    0.000600    0.001500    0.005800    0.006700    0.010200   
25%      0.013350    0.016450    0.018950    0.024375    0.038050    0.067025   
50%      0.022800    0.030800    0.034300    0.044050    0.062500    0.092150   
75%      0.035550    0.047950    0.057950    0.064500    0.100275    0.134125   
max      0.137100    0.233900    0.305900    0.426400    0.401000    0.382300   

              x_7         x_8         x_9        x_10  ...        x_51  \
count  208.000000  208.000000  208.000000  208.000000  ...  208.000000   
mean     0.121747    0.134799    0.178003    0.208259  ...    0.016069   
std      0.061788    0.085152    0.118387    0.134416  ...    0.012008   
min      0.003300    0.005500    0.007500    0.011300  ...    0.000000   
25%      0.080900    0.080425    0.097025    0.111275  ...    0.008425   
50%      0.106950    0.112100    0.152250    0.182400  ...    0.013900   
75%      0.154000    0.169600    0.233425    0.268700  ...    0.020825   
max      0.372900    0.459000    0.682800    0.710600  ...    0.100400   

             x_52        x_53        x_54        x_55        x_56        x_57  \
count  208.000000  208.000000  208.000000  208.000000  208.000000  208.000000   
mean     0.013420    0.010709    0.010941    0.009290    0.008222    0.007820   
std      0.009634    0.007060    0.007301    0.007088    0.005736    0.005785   
min      0.000800    0.000500    0.001000    0.000600    0.000400    0.000300   
25%      0.007275    0.005075    0.005375    0.004150    0.004400    0.003700   
50%      0.011400    0.009550    0.009300    0.007500    0.006850    0.005950   
75%      0.016725    0.014900    0.014500    0.012100    0.010575    0.010425   
max      0.070900    0.039000    0.035200    0.044700    0.039400    0.035500   

             x_58        x_59        x_60  
count  208.000000  208.000000  208.000000  
mean     0.007949    0.007941    0.006507  
std      0.006470    0.006181    0.005031  
min      0.000300    0.000100    0.000600  
25%      0.003600    0.003675    0.003100  
50%      0.005800    0.006400    0.005300  
75%      0.010350    0.010325    0.008525  
max      0.044000    0.036400    0.043900  

[8 rows x 60 columns]

In [7]:
data.shape

(208, 61)

In [8]:
data.isnull().sum()

x_1     0
x_2     0
x_3     0
x_4     0
x_5     0
       ..
x_57    0
x_58    0
x_59    0
x_60    0
Y       0
Length: 61, dtype: int64

In [11]:
data.duplicated().sum()

np.int64(0)

In [12]:
from sklearn.preprocessing import MinMaxScaler
X=data.drop('Y',axis=1)
y=data['Y']
scaler=MinMaxScaler()
X_scaled=scaler.fit_transform(X)

In [16]:
from sklearn.preprocessing import LabelEncoder
encoder = LabelEncoder()
y_encoded= encoder.fit_transform(y)

In [17]:
y_encoded

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

array([[0.13643068, 0.15645092, 0.13567674, ..., 0.18535469, 0.24517906,
        0.06004619],
       [0.32300885, 0.22160309, 0.27201051, ..., 0.10526316, 0.14049587,
        0.08775982],
       [0.18215339, 0.24689241, 0.35611038, ..., 0.36842105, 0.25895317,
        0.16628176],
       ...,
       [0.37389381, 0.18474068, 0.05420499, ..., 0.30892449, 0.20936639,
        0.05773672],
       [0.21238938, 0.14873553, 0.15604468, ..., 0.17391304, 0.09641873,
        0.09699769],
       [0.18067847, 0.15302186, 0.03975033, ..., 0.07551487, 0.16528926,
        0.2517321 ]])

2. Model Implementation
●	Construct a basic ANN model using your chosen high-level neural network library. Ensure your model includes at least one hidden layer.
●	Divide the dataset into training and test sets.
●	Train your model on the training set and then use it to make predictions on the test set.


In [22]:
#train test split
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X_scaled,y_encoded,test_size=0.2,random_state=42,stratify=y_encoded)


In [23]:
print(X_train.shape,X_test.shape)

(166, 60) (42, 60)


In [24]:
#model building
import tensorflow as tf
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense

In [26]:
num_classes=len(encoder.classes_)

In [27]:
num_classes

2

In [30]:
model = Sequential([
    Dense(128, activation='relu', input_shape=(60,)),   # 1 hidden layer
    Dense(64, activation='relu'),                       # Optional 2nd hidden layer
    Dense(num_classes, activation='softmax')            # Output layer
])


In [31]:
model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

In [32]:
model.summary()

Model: "sequential_2"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_6 (Dense)                 │ (None, 128)            │         7,808 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_7 (Dense)                 │ (None, 64)             │         8,256 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_8 (Dense)                 │ (None, 2)              │           130 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 16,194 (63.26 KB)

 Trainable params: 16,194 (63.26 KB)

 Non-trainable params: 0 (0.00 B)

In [33]:
#fitting the model
model.fit(
    X_train, y_train,
    validation_split=0.2,
    epochs=20,
    batch_size=32,
    verbose=1
)


Epoch 1/20
5/5 ━━━━━━━━━━━━━━━━━━━━ 1s 63ms/step - accuracy: 0.5627 - loss: 0.6792 - val_accuracy: 0.3824 - val_loss: 0.7298
Epoch 2/20
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 19ms/step - accuracy: 0.6080 - loss: 0.6325 - val_accuracy: 0.5882 - val_loss: 0.6555
Epoch 3/20
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 19ms/step - accuracy: 0.7395 - loss: 0.6024 - val_accuracy: 0.8824 - val_loss: 0.6052
Epoch 4/20
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 20ms/step - accuracy: 0.7543 - loss: 0.5966 - val_accuracy: 0.8824 - val_loss: 0.5818
Epoch 5/20
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 19ms/step - accuracy: 0.7806 - loss: 0.5851 - val_accuracy: 0.6765 - val_loss: 0.5821
Epoch 6/20
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 18ms/step - accuracy: 0.7252 - loss: 0.5538 - val_accuracy: 0.5882 - val_loss: 0.6164
Epoch 7/20
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 19ms/step - accuracy: 0.6921 - loss: 0.5381 - val_accuracy: 0.5882 - val_loss: 0.6125
Epoch 8/20
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 21ms/step - accuracy: 0.7427 - loss: 0.4993 - val_accuracy: 0.7059 - val_loss: 0.5199


In [34]:
test_loss, test_accuracy = model.evaluate(X_test, y_test)
print("Test Accuracy:", test_accuracy)


2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 103ms/step - accuracy: 0.8527 - loss: 0.3697
Test Accuracy: 0.8571428656578064


In [35]:
y_pred = model.predict(X_test)
y_pred_classes = y_pred.argmax(axis=1)

print(y_pred_classes[:10])


2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 66ms/step
[1 0 1 0 0 0 0 1 0 0]


3. Hyperparameter Tuning
●	Modify various hyperparameters, such as the number of hidden layers, neurons per hidden layer, activation functions, and learning rate, to observe their impact on model performance.
●	Adopt a structured approach like grid search or random search for hyperparameter tuning, documenting your methodology thoroughly.


In [53]:
!pip install keras-tuner

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 129.4/129.4 kB 9.3 MB/s eta 0:00:00


In [54]:
import keras_tuner as kt
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras.optimizers import Adam, SGD

def build_model(hp):

    model = Sequential()

    # Choose number of hidden layers (1 to 3)
    for i in range(hp.Int("num_layers", 1, 3)):
        model.add(Dense(
            units=hp.Choice("units_"+str(i), [32, 64, 128, 256]),
            activation=hp.Choice("activation", ["relu", "tanh"])
        ))

        # optional dropout
        if hp.Boolean("dropout_"+str(i)):
            model.add(Dropout(0.3))

    # Output layer
    model.add(Dense(num_classes, activation="softmax"))

    # Optimizer & learning rate
    lr = hp.Choice("learning_rate", [1e-4, 3e-4, 1e-3, 3e-3])
    opt = Adam(learning_rate=lr)

    model.compile(
        optimizer=opt,
        loss="sparse_categorical_crossentropy",
        metrics=["accuracy"]
    )

    return model


In [55]:
tuner = kt.RandomSearch(
    build_model,
    objective="val_accuracy",
    max_trials=10,          # number of hyperparameter combinations
    executions_per_trial=1,
    directory="ann_tuning",
    project_name="alphabet_model"
)


In [56]:
tuner.search(
    X_train, y_train,
    epochs=20,
    validation_split=0.2,
    verbose=2
)


Trial 10 Complete [00h 00m 05s]
val_accuracy: 0.6176470518112183

Best val_accuracy So Far: 0.9411764740943909
Total elapsed time: 00h 00m 53s


In [57]:
best_model = tuner.get_best_models(num_models=1)[0]
best_model.summary()


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 256)            │        15,616 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout (Dropout)               │ (None, 256)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 32)             │         8,224 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 2)              │            66 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 23,906 (93.38 KB)

 Trainable params: 23,906 (93.38 KB)

 Non-trainable params: 0 (0.00 B)

In [58]:
test_loss, test_acc = best_model.evaluate(X_test, y_test)
print("Test Accuracy:", test_acc)


2/2 ━━━━━━━━━━━━━━━━━━━━ 1s 73ms/step - accuracy: 0.8051 - loss: 0.4872 
Test Accuracy: 0.7857142686843872


4. Evaluation
●	Employ suitable metrics such as accuracy, precision, recall, and F1-score to evaluate your model's performance.
●	Discuss the performance differences between the model with default hyperparameters and the tuned model, emphasizing the effects of hyperparameter tuning


In [61]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, classification_report, confusion_matrix

# Predictions
y_pred_base = model.predict(X_test)
y_pred_base = y_pred_base.argmax(axis=1)

print("Baseline Model Classification Report:")
print(classification_report(y_test, y_pred_base, target_names=encoder.classes_))

print("Baseline Confusion Matrix:")
print(confusion_matrix(y_test, y_pred_base))


2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 72ms/step
Baseline Model Classification Report:
              precision    recall  f1-score   support

           M       0.81      0.95      0.88        22
           R       0.94      0.75      0.83        20

    accuracy                           0.86        42
   macro avg       0.87      0.85      0.85        42
weighted avg       0.87      0.86      0.86        42

Baseline Confusion Matrix:
[[21  1]
 [ 5 15]]


In [63]:
# Predictions from tuned model
y_pred_tuned = best_model.predict(X_test)
y_pred_tuned = y_pred_tuned.argmax(axis=1)

print("Tuned Model Classification Report:")
print(classification_report(y_test, y_pred_tuned, target_names=encoder.classes_))

print("Tuned Confusion Matrix:")
print(confusion_matrix(y_test, y_pred_tuned))


2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 33ms/step
Tuned Model Classification Report:
              precision    recall  f1-score   support

           M       0.84      0.73      0.78        22
           R       0.74      0.85      0.79        20

    accuracy                           0.79        42
   macro avg       0.79      0.79      0.79        42
weighted avg       0.79      0.79      0.79        42

Tuned Confusion Matrix:
[[16  6]
 [ 3 17]]


Conclusion
as from the given dataset we perform the data exploration and preprocessing then we check the missing values and duplicates data then we normalize the data by using min max scaling technique and for categorical column wr apply label encoding. then we split the data into train and testing by using train_test_split and then we build an ANN Base model through input layer and one hidden layer and an output layer by using keras we build the model.
after build the base model we got an accuracy of 0.85 and a good optimizer we used Adam, reasonable learning rate 0.0001 , simple architecture that avoids overfitting.our base model achieved higher accuracy, it indicates that the model configuration was already well-suited to the dataset.

After base model we tuned our model by using random search, it explores the hyperparameter space with randomly

Many combinations may not work well:

1.Too high or too low learning rate

2.Too few or too many neurons

3.Wrong activation functions

4.Too deep networks → overfitting

by using our tuned model we got the model accuracy of 0.78
this are the reasons behind the less accuracy then the Base ANN  model.